helper/metricsutil/gauge_process.go

```diff
@@ -0,0 +1,236 @@
+package metricsutil
+
+import (
+	"context"
+	"math/rand"
+	"sort"
+	"time"
+
+	log "github.com/hashicorp/go-hclog"
+)
+
+// This interface allows unit tests to substitute in a simulated clock.
+type clock interface {
+	Now() time.Time
+	NewTicker(time.Duration) *time.Ticker
+}
+
+type defaultClock struct {
+}
+
+func (_ defaultClock) Now() time.Time {
+	return time.Now()
+}
+
+func (_ defaultClock) NewTicker(d time.Duration) *time.Ticker {
+	return time.NewTicker(d)
+}
+
+// GaugeLabelValues is one gauge in a set sharing a single key, that
+// are measured in a batch.
+type GaugeLabelValues struct {
+	Labels []Label
+	Value  float32
+}
+
+// GaugeCollector is a callback function that returns an unfiltered
+// set of label-value pairs. It may be cancelled if it takes too long.
+type GaugeCollector = func(context.Context) ([]GaugeLabelValues, error)
+
+// collectionBound is a hard limit on how long a collection process
+// may take, as a fraction of the current interval.
+const collectionBound = 0.02
+
+// collectionTarget is a soft limit; if exceeded, the collection interval
+// will be doubled.
+const collectionTarget = 0.01
+
+// A GaugeCollectionProcess is responsible for one particular gauge metric.
+// It handles a delay on initial startup; limiting the cardinality; and
+// exponential backoff on the requested interval.
+type GaugeCollectionProcess struct {
+	Stop    chan struct{}
+	stopped chan struct{}
+
+	// gauge name
+	key []string
+	// labels to use when reporting
+	labels []Label
+
+	// callback function
+	collector GaugeCollector
+
+	// destination for metrics
+	sink   *ClusterMetricSink
+	logger log.Logger
+
+	// time between collections
+	originalInterval time.Duration
+	currentInterval  time.Duration
+	ticker           *time.Ticker
+
+	// time source
+	clock clock
+}
+
+// NewGaugeCollectionProcess creates a new collection process for the callback
+// function given as an argument, and starts it running.
+// A label should be provided for metrics *about* this collection process.
+//
+// The Run() method must be called to start the process.
+func (m *ClusterMetricSink) NewGaugeCollectionProcess(
+	key []string,
+	id []Label,
+	collector GaugeCollector,
+	logger log.Logger,
+) (*GaugeCollectionProcess, error) {
+	return m.newGcpWithClock(
+		key,
+		id,
+		collector,
+		logger,
+		defaultClock{},
+	)
+}
+
+// test version allows an alternative clock implementation
+func (m *ClusterMetricSink) newGcpWithClock(
```

Comment:
I would rather see the full name than the abbreviation (newGaugeCollectionProcessWithClock), even if it is verbose.  Not least because I keep thinking of Google Cloud Platform when I read this. :)

Reply:
Changed.